### PR TITLE
Robustify multi-threading primitives

### LINKED ISSUE #71216: Deadlock on exit on Linux in projects using Thread/Semaphore in an autoload (e.g. for background resource loading)
### Godot version

3.6.beta (1426cd3), 4.0.beta (9c02bf1)

### System information

Mageia 9, Linux / X11

### Issue description

This issue will need to be fleshed out further when the problem can be better narrowed down, but @jitspoe and I noticed a deadlock issue when closing some projects on Linux (which doesn't seem to be reproduced on Windows).

Both projects tested use Thread and Semaphore in an autoload to do background loading of resources, and when finalizing the SceneTree and destructing the Node, there seems to be a deadlock (waiting on the Semaphore while destructing the autoload's GDScriptInstance?).

The problem was observed in:

- @jitspoe's Fist of the Forgotten using 3.6.beta (custom fork). It can be reproduced with the [Steam demo](https://store.steampowered.com/app/1105470/Fist_of_the_Forgotten/), which apparently still includes enough debug symbols for GDB to give me this stacktrace:

```
Thread 1 "FistOfTheForgot" received signal SIGINT, Interrupt.
0x00007f71f41e157d in pthread_cond_destroy@@GLIBC_2.3.2 () from /lib64/libc.so.6
(gdb) bt
#0  0x00007f71f41e157d in pthread_cond_destroy@@GLIBC_2.3.2 () from /lib64/libc.so.6
#1  0x0000000001efb7de in _Semaphore::~_Semaphore() ()
#2  0x000000000049eb2e in Ref<Reference>::unref() ()
#3  0x0000000001d59da7 in Variant::clear() ()
#4  0x00000000004da105 in CowData<Variant>::_unref(void*) ()
#5  0x00000000006d469e in GDScriptInstance::~GDScriptInstance() ()
#6  0x0000000001ef87de in Object::~Object() ()
#7  0x0000000000c2c75a in void memdelete<Node>(Node*) ()
#8  0x0000000000c8047f in Node::_notification(int) ()
#9  0x0000000000c81037 in Viewport::_notificationv(int, bool) ()
#10 0x0000000000c5774f in SceneTree::finish() ()
#11 0x00000000004452aa in main ()
```

- @BeyondXinXin's Godot 4 "God-Game" which has source code included in #71205. Beside the bug in that issue, it exhibits the same deadlock on exit on Linux as Fist of the Forgotten. To reproduce it (without triggering #71205, run the game, press the button to start the level, and quit before trying to move the character (which is what would trigger #71205 in current `master`, at least until that regression is fixed).

```
(gdb) bt
#0  0x00007ffff7d6757d in pthread_cond_destroy@@GLIBC_2.3.2 () from /lib64/libc.so.6
#1  0x000000000592a71a in Semaphore::~Semaphore (this=0x7fffac0036c8, __in_chrg=<optimized out>) at ./core/os/semaphore.h:40
#2  0x0000000009debcaa in core_bind::Semaphore::~Semaphore (this=0x7fffac003530, __in_chrg=<optimized out>) at ./core/core_bind.h:367
#3  0x000000000599b537 in memdelete<RefCounted> (p_class=0x7fffac003530) at ./core/os/memory.h:109
#4  0x000000000a043ab5 in Variant::_clear_internal (this=0xbb11650) at core/variant/variant.cpp:1354
#5  0x000000000536cebf in Variant::clear (this=0xbb11650) at ./core/variant/variant.h:302
#6  0x000000000536cf2c in Variant::~Variant (this=0xbb11650, __in_chrg=<optimized out>) at ./core/variant/variant.h:778
#7  0x00000000053a5e56 in CowData<Variant>::_unref (this=0xeed03f0, p_data=0xbb115f0) at ./core/templates/cowdata.h:213
#8  0x00000000053a1aee in CowData<Variant>::~CowData (this=0xeed03f0, __in_chrg=<optimized out>) at ./core/templates/cowdata.h:412
#9  0x000000000539ce74 in Vector<Variant>::~Vector (this=0xeed03e8, __in_chrg=<optimized out>) at ./core/templates/vector.h:287
#10 0x0000000005d90bab in GDScriptInstance::~GDScriptInstance (this=0xeed0390, __in_chrg=<optimized out>) at modules/gdscript/gdscript.cpp:1955
#11 0x000000000a374012 in memdelete<ScriptInstance> (p_class=0xeed0390) at ./core/os/memory.h:109
#12 0x000000000a3715f1 in Object::~Object (this=0xd028db0, __in_chrg=<optimized out>) at core/object/object.cpp:1769
#13 0x0000000008314d31 in Node::~Node (this=0xd028db0, __in_chrg=<optimized out>) at scene/main/node.cpp:3036
#14 0x000000000596192f in memdelete<Node> (p_class=0xd028db0) at ./core/os/memory.h:109
#15 0x00000000082fc83a in Node::_notification (this=0xed094b0, p_notification=1) at scene/main/node.cpp:169
#16 0x0000000005b0f91c in Node::_notificationv (this=0xed094b0, p_notification=1, p_reversed=true) at ./scene/main/node.h:45
#17 0x0000000006561a29 in Viewport::_notificationv (this=0xed094b0, p_notification=1, p_reversed=true) at ./scene/main/viewport.h:89
#18 0x000000000656231b in Window::_notificationv (this=0xed094b0, p_notification=1, p_reversed=true) at ./scene/main/window.h:44
#19 0x000000000a368155 in Object::notification (this=0xed094b0, p_notification=1, p_reversed=true) at core/object/object.cpp:790
#20 0x000000000a365ae7 in Object::_predelete (this=0xed094b0) at core/object/object.cpp:196
#21 0x000000000a371b15 in predelete_handler (p_object=0xed094b0) at core/object/object.cpp:1823
#22 0x0000000008343c09 in memdelete<Window> (p_class=0xed094b0) at ./core/os/memory.h:105
#23 0x00000000083355a1 in SceneTree::finalize (this=0xed76090) at scene/main/scene_tree.cpp:584
#24 0x0000000005374716 in OS_LinuxBSD::run (this=0x7fffffffd1d0) at platform/linuxbsd/os_linuxbsd.cpp:883
#25 0x000000000536c140 in main (argc=1, argv=0x7fffffffd738) at platform/linuxbsd/godot_linuxbsd.cpp:73
```

Neither project is minimal, but I figured I'd open this issue to start the debugging process and put it on @RandomShaper's radar :)

It might be a misuse of Thread/Semaphore in both projects, but given that it seems to deadlock only on Linux and not on Windows, it should probably still be considered a bug as it's hard to debug for Windows users and harms cross-platform deployment.

### Steps to reproduce

TBD, see above for now.

### Minimal reproduction project

I managed to extract a MRP from the project in #71205:
[linux_semaphore_deadlock.zip](https://github.com/godotengine/godot/files/10392308/linux_semaphore_deadlock.zip)

On Linux, it deadlocks, while on Windows, it closes fine but with these errors:
```
ERROR: Condition "_first != nullptr" is true.
   at: ~List (./core/templates/self_list.h:106)
WARNING: ObjectDB instances leaked at exit (run with --verbose for details).
     at: cleanup (core/object/object.cpp:1980)
ERROR: Resources still in use at exit (run with --verbose for details).
   at: clear (core/io/resource.cpp:473)
```

### PR DESCRIPTION
**Pull request by RandomShaper**
This PR does the following:
- Fixes the issue that `Thread` objects were leaked to the OS if not joined (i.e., in those cases there would be no possibility for the warning about calling `wait_to_finish()` to be printed; therefore, pretty much hiding the issue to the user).
- In debug builds, detects wrong cleanup of `Semaphore` (long explanation in the code comments; TL;DR is _Don't exit with semaphores being awaited_) and avoids the undefined behavior. I have tried to add similar protection to `Mutex`, but there are no reasonable ways to detect misuse. In any case, now we can finally get the warning for non-joined threads, most (all?) cases of misuse would cause some kind of admonishment.
- Documents the details about proper usage of the related classes, to prevent users from running into undefined behavior, including those we can detect and warn about or countermeasure and those that don't.

Fixes #71216.

**UPDATE:** Back-ported to 3.x in #72251.

### CODE REVIEW

File: core/os/thread.cpp [hunk 1]
```diff
@@ -113,7 +113,9 @@ Thread::Thread() {
 Thread::~Thread() {
 	if (id != _thread_id_hash(std::thread::id())) {
 #ifdef DEBUG_ENABLED
-		WARN_PRINT("A Thread object has been destroyed without wait_to_finish() having been called on it. Please do so to ensure correct cleanup of the thread.");
+		WARN_PRINT(
```
Comment (RandomShaper) [inline core/os/thread.cpp:116]: I've just reworded this one to be consistent with the warning newly added to `Semaphore`.

**Comment by YuriSizov** [comment]
Given 3.x is also affected, this should be safe to merge for 4.0.x when reduz has time to review it.

**Comment by RandomShaper** [comment]
Pushed just fixing a couple of mistakes in the docs.

**Comment by RandomShaper** [comment]
Rebased on top of 4.1.

File: core/os/semaphore.h [hunk 2]
```diff
@@ -42,6 +45,9 @@ class Semaphore {
 	mutable std::mutex mutex;
 	mutable std::condition_variable condition;
 	mutable uint32_t count = 0; // Initialized as locked.
+#ifdef DEBUG_ENABLED
+	mutable uint32_t awaiters = 0;
```
Comment (reduz) [inline core/os/semaphore.h:49]: Should this not be atomic?

Comment (RandomShaper) [reply core/os/semaphore.h:49]: It's always protected by the mutex.

**Review by reduz — APPROVED** [review]

File: core/os/semaphore.h (hunk 2 above)
Comment (reduz) [reply core/os/semaphore.h:49]: ah I missed the mutex, nevermind.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**